### PR TITLE
closes #357 Lesson Info for the admin lessons page

### LINKED ISSUE #357: Admin Lessons Info
The Admin Lessons page needs to display information about a lesson when a `lesson title` on the sidebar of the lessons page is clicked on.

### PR DESCRIPTION
**Pull request by anthonykhoa**
The Admin Lessons page needs to display information about a lesson when a lesson title on the sidebar of the lessons page is clicked on. It also needs to be able to create a new lesson.

This PR will:
* Display lesson title when person clicks a `lesson title in the sidebar`
* Ability to create new Lesson by clicking on `create new lesson` button located in the sidebar

# Lesson Info
Screen Shot 2020-08-01 at 7 10 17 PM

# Create New Lesson
Screen Shot 2020-08-01 at 7 10 35 PM

### CODE REVIEW

File: helpers/admin/lessonHelpers.ts [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+    }
+  }
+  return errorSeen
+}
+
+//checks for error for each element in the `inputvalues` array
+export const checkForAllErrors = (options: any) => {
+  let error = false
+  options.forEach((option: any) => {
+    if (checkForErrors(option)) error = true
```
Comment (Oroul) [inline helpers/admin/lessonHelpers.ts:61]: ``

File: components/admin/lessons/AdminLessonInfo.tsx [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+  }
+
+  const handleChange = (value: string, propertyIndex: number) => {
+    const newLessonInfo = [...lessonInfo]
+    newLessonInfo[propertyIndex].value = value
+    checkForErrors(newLessonInfo[propertyIndex])
+    setLessonInfo(newLessonInfo)
+  }
+  return (
+    <div style={{ textAlign: 'center', marginBottom: 20 }} className=" col-8">
```
Comment (SahilKalra98) [inline components/admin/lessons/AdminLessonInfo.tsx:165]: Not a big issue, ```className=" col-8"``` has an extra space

Comment (anthonykhoa) [reply components/admin/lessons/AdminLessonInfo.tsx:165]: thank you. I will fix it right now!

File: helpers/admin/lessonHelpers.ts [hunk 3]
```diff
@@ -0,0 +1,64 @@
+// creates usuable array from graphql data to use as a prop when using FormCard component
+export const inputValues = (options: any, deleteProp?: string) => {
+  deleteProp && options.hasOwnProperty(deleteProp) && delete options[deleteProp]
+  options.hasOwnProperty('__typename') && delete options['__typename']
+  const keys = Object.keys(options)
+  const res = keys.reduce((acc: any, type: any) => {
```
Comment (songz) [inline helpers/admin/lessonHelpers.ts:6]: did you mean `map`?

Comment (anthonykhoa) [reply helpers/admin/lessonHelpers.ts]: thank you yes

File: helpers/admin/lessonHelpers.ts [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+      title: type,
+      value,
+      type: type === 'description' ? 'MD_INPUT' : 'TEXT_AREA'
+    }
+  })
+  return res
+}
+
+// turns the array used in FormCard component into an object, to be used when making mutation requests
+export const outputValues = (options: any) => {
```
Comment (songz) [inline helpers/admin/lessonHelpers.ts:18]: `outputValues` -> `makeLessonVariable`

File: components/admin/lessons/AdminLessonInfo.tsx [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+      title,
+      description,
+      docUrl,
+      githubUrl,
+      videoUrl,
+      order,
+      chatUrl
+    } = outputValues(lessonInfo)
+    try {
+      await alterLesson({
```
Comment (songz) [inline components/admin/lessons/AdminLessonInfo.tsx:59]: `alterLesson( makeLessonVariable(lessonInfo) )`

File: components/admin/lessons/AdminLessonInfo.tsx [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+      title,
+      description,
+      docUrl,
+      githubUrl,
+      videoUrl,
+      order,
+      chatUrl
+    } = outputValues(lessonInfo)
+    try {
+      await createLesson({
```
Comment (songz) [inline components/admin/lessons/AdminLessonInfo.tsx:141]: `createLesson( makeLessonVariable(lessonInfo) )`

File: components/admin/lessons/AdminLessonInfo.tsx [hunk 7]
```diff
@@ -0,0 +1,197 @@
...
+  setLessons: React.Dispatch<React.SetStateAction<Lesson[] | null>>
+  lesson: Lesson | undefined
+}
+
+type NewLessonProps = {
+  setLessons: React.Dispatch<React.SetStateAction<Lesson[] | null>>
+}
+
+// Creates card for a lessons's information to update
+const LessonBase: React.FC<LessonBaseProps> = ({ setLessons, lesson }) => {
```
Comment (songz) [inline components/admin/lessons/AdminLessonInfo.tsx:31]: `EditLesson`

File: helpers/admin/lessonHelpers.ts [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+    }
+  } else if (title === 'title' || title === 'description') {
+    if (!value) {
+      option.error = 'Required'
+      errorSeen = true
+    } else {
+      option.hasOwnProperty('error') && delete option.error
+    }
+  }
+  return errorSeen
```
Comment (songz) [inline helpers/admin/lessonHelpers.ts:48]: your nested if conditions are confusing:

```
const required = ['title', 'description', 'order']
if ( required.includes(title) ) {
  option.error = 'Required'
  return true
}
if (title === 'order' && !value.match(....) ) {
  option.error = 'Numbers only'
  return true
}
...
delete option.error
return false
```

File: helpers/admin/lessonHelpers.ts [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+}
+
+//checks for error for each element in the `inputvalues` array
+export const checkForAllErrors = (options: any) => {
+  let error = false
+  options.forEach((option: any) => {
```
Comment (songz) [inline helpers/admin/lessonHelpers.ts:54]:
```
const hasError = options.find(...)
return !!hasError
```

File: helpers/admin/lessonHelpers.ts [hunk 10]
```diff
@@ -0,0 +1,58 @@
+// creates usuable array from graphql data to use as a prop when using FormCard component
+export const inputValues = (options: any, deleteProp?: string) => {
```
Comment (songz) [inline helpers/admin/lessonHelpers.ts:2]: `getPropertyArr = (options: any, deleteProps?: array of strings`

File: helpers/admin/lessonHelpers.ts [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+  deleteProp && options.hasOwnProperty(deleteProp) && delete options[deleteProp]
```
Comment (songz) [inline helpers/admin/lessonHelpers.ts:3]:
```
(deleteProp || []).forEach( prop => delete options[prop] )
```

File: helpers/admin/lessonHelpers.ts [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+  options.hasOwnProperty('__typename') && delete options['__typename']
```
Comment (songz) [inline helpers/admin/lessonHelpers.ts:4]: `delete options.__typename`

**Pull request merged**